### ISSUE TITLE
reset() cannot be used in Effect Hooks

### ISSUE THREAD

**Issue by sheraz-sarwar** [bluebill1049/react-hook-form#126]
**Describe the bug**
To preface, this may be intentional behaviour although if it is, I don't see mention of it anywhere in the docs.

I've written a form which populates its fields using the props passed to the component. Naturally, these props can change following user interaction. I would hope to be able to easily reinitialise the form values based on the new props passed to the form component.

To begin, inputs should be uncontrolled and therefore take the `defaultValue` attribute, like so.

```
<input
  type="text"
  name="name"
  defaultValue={name}
  ref={register}
/>
```

Then, to reset the form when the `name` prop changes, I do the following.

```
const { register, handleSubmit, reset } = useForm();

useEffect(() => {
  reset();
}, [name]);
```

This works but it violates the `react-hooks/exhaustive-deps` eslint rule since `reset` is considered a missing dependency. However, adding `reset` as a dependency creates an infinite loop whereby:

1. `name` is changed therefore the effect is triggered
2. The effect calls `reset` which changes the `useForm()` methods, including `reset` itself
3. `reset` is changed therefore the effect is triggered

As per [the Hooks documentation](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies), it is generally _not_ safe to omit functions from the list of dependencies and "the recommended fix is to move that function inside of your effect". In this case, this isn't possible either. `reset` is returned by a hook, which cannot be called inside a callback.

In summary:
1. `reset` cannot be defined within an Effect Hook
2. if `reset` is defined outside of an Effect Hook then called within it:
  a. an infinite loop is created if `reset` is declared as a dependency
  b. the `exhaustive-deps` rule is violated if it is not declared as a dependency

I'm not sure if there's an alternative option that I'm not seeing here.

**Steps to reproduce**
I've recreated the issue using a very simple example at the following link: https://codesandbox.io/s/react-hook-form-reset-form-r35fi?fontsize=14

Feel free to amend and suggest an alternative way of repopulating form fields from changed props. Please note, I've tried using `setValue` instead and hit almost exactly the same problems.

**Comment by bluebill1049** [comment]
thanks very much for the detailed issue report. really helpful 👍

**Comment by bluebill1049** [comment]
can you test out this beta version? react-hook-form@3.18.5-beta.1

see if it's fixing the issue.

**Comment by sheraz-sarwar** [comment]
Just tested this in our application using the above version: declaring `reset` as a dependency in the Effect Hook no longer triggers a loop. Thanks for looking at this so quickly!

**Comment by bluebill1049** [comment]
my pleasure @sheraz-sarwar publishing now `react-hook-form@3.18.5` 🙏 please consider supporting this repo by leaving a star and share in the community. thanks

**Comment by bluebill1049** [comment]
@sheraz-sarwar feel free to close this issue when you happy with it 👍

**Comment by sheraz-sarwar** [comment]
Closed, starred, and shared. Thanks again.

**Comment by bluebill1049** [comment]
Thank you so much 😊
